source/channel_bootstrap.c

```diff
@@ -108,6 +185,116 @@ struct client_connection_args {
+    int err_code,
+    void *user_data) {
+    struct client_connection_args *connection_args = user_data;
+
+    if (connection_args->channel_data.user_on_negotiation_result) {
+        connection_args->channel_data.user_on_negotiation_result(
+            handler, slot, err_code, connection_args->channel_data.tls_user_data);
+    }
+
+    if (!err_code) {
+        connection_args->setup_callback(
+            connection_args->bootstrap,
+            AWS_OP_SUCCESS,
+            &connection_args->channel_data.channel,
+            connection_args->user_data);
+    } else {
+        /* the tls handler is responsible for calling shutdown when this fails. */
+        connection_args->setup_callback(connection_args->bootstrap, err_code, NULL, connection_args->user_data);
+    }
+}
+
+/* in the context of a channel bootstrap, we don't care about these, but since we're hooking into these APIs we have to
+ * provide a proxy for the user actually receiving their callbacks. */
+static void s_tls_client_on_data_read(
+    struct aws_channel_handler *handler,
+    struct aws_channel_slot *slot,
+    struct aws_byte_buf *buffer,
+    void *user_data) {
+    struct client_connection_args *connection_args = user_data;
+
+    if (connection_args->channel_data.user_on_data_read) {
+        connection_args->channel_data.user_on_data_read(
+            handler, slot, buffer, connection_args->channel_data.tls_user_data);
+    }
+}
+
+/* in the context of a channel bootstrap, we don't care about these, but since we're hooking into these APIs we have to
+ * provide a proxy for the user actually receiving their callbacks. */
+static void s_tls_client_on_error(
+    struct aws_channel_handler *handler,
+    struct aws_channel_slot *slot,
+    int err,
+    const char *message,
+    void *user_data) {
+    struct client_connection_args *connection_args = user_data;
+
+    if (connection_args->channel_data.user_on_error) {
+        connection_args->channel_data.user_on_error(
+            handler, slot, err, message, connection_args->channel_data.tls_user_data);
+    }
+}
+
+static inline int s_setup_client_tls(struct client_connection_args *connection_args, struct aws_channel *channel) {
+    struct aws_channel_slot *tls_slot = aws_channel_slot_new(channel);
+
+    /* as far as cleanup goes, since this stuff is being added to a channel, the caller will free this memory
+       when they clean up the channel. */
+    if (!tls_slot) {
+        return AWS_OP_ERR;
+    }
+
+    struct aws_channel_handler *tls_handler = aws_tls_client_handler_new(
+        connection_args->bootstrap->allocator,
+        connection_args->bootstrap->tls_ctx,
+        &connection_args->channel_data.tls_options,
+        tls_slot);
+
+    if (!tls_handler) {
+        aws_mem_release(connection_args->bootstrap->allocator, (void *)tls_slot);
+        return AWS_OP_ERR;
+    }
+
+    aws_channel_slot_insert_end(channel, tls_slot);
+    if (aws_channel_slot_set_handler(tls_slot, tls_handler)) {
```

Comment:
This inconsistency in freeing things makes me uneasy, because ownership is managed *somewhere else*, except when it's not for some errors. Any reason not to clean up here if possible? Otherwise, this function isn't particularly atomic, which feels like a good source for state bugs or leaks later.

Reply:
Also, this would be clearer if it were != AWS_OP_SUCCESS

Reply:
so the readability part.... done

On the first comment, yeah, that's good feedback. We should be consistent on error means I didn't take ownership of your memory thing. But.... in this case.... that's not the case, because it's the window update that failed and the handler was set. Let's circle back on this after this PR.